daemon: create globals directory when other directories are created

ianvernon opened the pull request:
Based on https://github.com/cilium/cilium/pull/9183

This reduces the creation of directories from being spread across different
parts of the code base.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9184)
<!-- Reviewable:end -->

Commit:
daemon: create globals directory when other directories are created

This reduces the creation of directories from being spread across different
parts of the code base.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/daemon.go:
```diff
@@ -491,6 +486,7 @@ func NewDaemon(dp datapath.Datapath, iptablesManager rulesManager) (*Daemon, *en
 
 	bootstrapStats.bpfBase.Start()
 	err = d.init()
+
```
Seems unrelated.

aanm commented on daemon/daemon.go:
```diff
@@ -205,11 +205,6 @@ func (d *Daemon) GetCompilationLock() *lock.RWMutex {
-		log.WithError(err).WithField(logfields.Path, globalsDir).Fatal("Could not create runtime directory")
-	}
-
 	if err := os.Chdir(option.Config.StateDir); err != nil {
```
Shouldn't this be moved as well?

aanm commented on daemon/daemon_main.go:
```diff
@@ -841,6 +841,12 @@ func initEnv(cmd *cobra.Command) {
 	if err := os.MkdirAll(option.Config.LibDir, defaults.RuntimePathRights); err != nil {
 		scopedLog.WithError(err).Fatal("Could not create library directory")
 	}
+
```
I would suggest to have a dedicated function for directories stuff

aanm commented on daemon/daemon_test.go:
```diff
@@ -86,6 +86,9 @@ func TestMain(m *testing.M) {
 		panic("TempDir() failed.")
 	}
+
+	defer os.RemoveAll(tempRunDir)
```
was `option.Config.RunDir` set with `tempRunDir`?

aanm commented on daemon/daemon_test.go:
```diff
@@ -149,10 +152,6 @@ func (ds *DaemonSuite) SetUpTest(c *C) {
 	ds.d.endpointManager.RemoveAll()
 
-	if ds.d != nil {
-		os.RemoveAll(option.Config.RunDir)
```
Won't this leave leftover files from the tests?

Commit:
daemon: move changing to `StateDir`

Move this to be more logically colocated to where the directories are created.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: factor out directory creation into function

Signed-off by: Ian Vernon <ian@cilium.io>
